src/qibo/models.py

```diff
@@ -429,3 +431,161 @@ def _loss(params, qaoa, hamiltonian):
                                                              compile=compile, processes=processes)
         self.set_parameters(parameters)
         return result, parameters, extra
+
+
+class Grover:
+    """Model that performs Grover's algorithm.
+
+    # TODO: Add a few more details on Grover's algorithm and/or reference.
+
+    Args:
+        oracle (:class:`qibo.core.circuit.Circuit`): quantum circuit that flips
+            the sign using a Grover ancilla initialized with -X-H-
+            and expected to have the total size of the circuit.
+        initial_state (:class:`qibo.core.circuit.Circuit`): quantum circuit
+            that initializes the state. Leave empty if |000..00>
+        superposition (:class:`qibo.core.circuit.Circuit`): quantum circuit that
+            takes an initial state to a superposition. Expected to use the first
+            set of qubits to store the relevant superposition.
+        sup_qubits (int): number of qubits that store the relevant superposition.
+            Leave empty if superposition does not use ancillas.
+        sup_size (int): how many states are in a superposition.
+            Leave empty if its an equal superposition of quantum states.
+        num_sol (int): number of expected solutions. Needed for normal Grover.
+            Leave empty for iterative version.
+        check (function): function that returns True if the solution has been
+            found. Required of iterative approach.
+            First argument should be the bitstring to check.
+        check_args (tuple): arguments needed for the check function.
+            The found bitstring not included.
+    """
+    def __init__(self, oracle, superposition=None, initial_state=None, sup_qubits=None, sup_size=None, num_sol=None, check=None, check_args=()):
+        self.oracle = oracle
+        self.initial_state = initial_state
```

Comment:
I think it is probably worth add a dimension check taking the nqubits from the oracle and then checking the length of the initial state.

Reply:
Have reworded `initial_state` into `initial_state_circuit` as I was looking for a quantum circuit that prepares the state.